### PR TITLE
Warn for #[unstable] on trait impls when it has no effect.

### PR DESCRIPTION
**Pull request by m-ou-se**
Earlier today I sent a PR with an `#[unstable]` attribute on a trait `impl`, but was informed that this attribute has no effect there. (comment: https://github.com/rust-lang/rust/pull/76525#issuecomment-689678895, issue: https://github.com/rust-lang/rust/issues/55436)

This PR adds a warning for this situation. Trait `impl` blocks with `#[unstable]` where both the type and the trait are stable will result in a warning:

```
warning: An `#[unstable]` annotation here has no effect. See issue #55436 <https://github.com/rust-lang/rust/issues/55436> for more information.
   --> library/std/src/panic.rs:235:1
    |
235 | #[unstable(feature = "integer_atomics", issue = "32976")]
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

---

It detects three problems in the existing code:

1. A few `RefUnwindSafe` implementations for the atomic integer types in `library/std/src/panic.rs`. Example:
https://github.com/rust-lang/rust/blob/d92155bf6ae0b7d79fc83cbeeb0cc0c765353471/library/std/src/panic.rs#L235-L236
2. An implementation of `Error` for `LayoutErr` in `library/std/srd/error.rs`:
https://github.com/rust-lang/rust/blob/d92155bf6ae0b7d79fc83cbeeb0cc0c765353471/library/std/src/error.rs#L392-L397
3. `From` implementations for `Waker` and `RawWaker` in `library/alloc/src/task.rs`. Example:
https://github.com/rust-lang/rust/blob/d92155bf6ae0b7d79fc83cbeeb0cc0c765353471/library/alloc/src/task.rs#L36-L37

Case 3 interesting: It has a bound with an `#[unstable]` trait (`W: Wake`), so appears to have not much effect on stable code. It does however break similar blanket implementations. It would also have immediate effect if `Wake` was implemented for any stable type. (Which is not the case right now, but there are no warnings in place to prevent it.) Whether this case is a problem or not is not clear to me. If it isn't, adding a simple `c.visit_generics(..);` to this PR will stop the warning for this case.

### CODE REVIEW

**Comment by petrochenkov** [comment]
I this PR exists, then stability checking has been broken at some point.
Stability checking is supposed to either require the `#[unstable]` attribute (if the item it's applied to is reachable), or produce an error (the attribute is not applicable to this item).

...unless `#[unstable]` is required, but it was already inherited from a parent module, so writing it on the item itself is not necessary.

Anyway, stability attributes on impls are actually required if not inherited, and very much make sense at least for documentation, and there were always plans to implement stability checking for impls some day, but it's hard and not top priority.

**Comment by petrochenkov** [comment]
I don't think complicating stability checking is the right trade off here, it already accumulated quite a bit of technical debt over the years.

**Comment by m-ou-se** [comment]
`#[unstable]` for `impl Trait for Type` blocks has never worked. (https://github.com/rust-lang/rust/issues/55436)

It's easy to make the mistake of adding a `#[unstable]` trait impl that's effectively stable. This has happened by accident. It's also easy to forget to stabilize one when the relevant type or trait gets stabilized, as has happened for both `LayoutErr` and `RefUnwindSafe`. This warning prevents this. Seems well worth it.

**Comment by lcnr** [comment]
I have already both made this mistake myself and saw a few other PRs which did the same, so :+1: on adding a warning here.
The implementation also looks good to me.

@m-ou-se Considering that you mentioned that this detected 3 mistakes, how does this pass CI without having to fix any of them? Am I missing something here?

**Comment by m-ou-se** [comment]
> @m-ou-se Considering that you mentioned that this detected 3 mistakes, how does this pass CI without having to fix any of them? Am I missing something here?

They're just warnings right now. So they do appear in the output, but don't fail the CI:
image

They should be errors instead, like the other messages about stability attributes.

I'm not sure how to solve case 3 though.

**Comment by lcnr** [comment]
We can maybe change this to an unstable lint and explicitly allow it in this case :thinking:

**Comment by m-ou-se** [comment]
I turned them into errors.

The first case is fixed using
```rust
#[stable(feature = "alloc_layout", since = "1.28.0")]
```

The second case is fixed using
```rust
#[stable(feature = "integer_atomics_stable", since = "1.34.0")]
```

Might be good if someone can verify these. I copied the stability attribute from the types, and checked that the (stable) traits and these ('unstable') trait impls already existed at the time the types were stabilized.

The third case still needs a solution. A lint with `#[allow(..)]` sounds like it could work. Will look into that.

**Comment by m-ou-se** [comment]
Any idea where would be the best place to call `register_lints` for this lint? Nothing in `compiler/rustc_passes` registers any lints, but only uses bulitin lints. But adding this to the builtin lints doesn't seem like a good idea, as it has no use outside of the standard libraries.

**Comment by lcnr** [comment]
hmm, maybe adding it to `builtin_lints` and only running it if the `staged_api` feature is active?

**Comment by m-ou-se** [comment]
Added it to the builtin lints as a `rustc::` tool lint, just like the internal lints. That way it doesn't show up in `-W help` and an `#[allow()]` attribute for it doesn't trigger a unknown lint error (which happens when bootstrapping rustc+std, using the current beta).

Also added a test, as the 'test cases' in std are now fixed.

File: library/alloc/src/task.rs [hunk 1]
```diff
@@ -33,6 +33,7 @@ pub trait Wake {
     }
 }
 
+#[allow(rustc::ineffective_unstable_trait_impl)]
```
Comment (lcnr) [inline library/alloc/src/task.rs:36]: Can you add a short comment here explaining why this is ok?

Something like
```suggestion
// While this impl is automatically stable, it only applies to types implementing `Wake`,
// which is unstable and not implemented for any stable type.
#[allow(rustc::ineffective_unstable_trait_impl)]
```

**Comment by lcnr** [comment]
> an `#[allow()]` attribute for it doesn't trigger a unknown lint error

Does this mean that the following test compiles rn?
```rust
#![allow(rustc::ineffective_unstable_trait_impl)]…
fn main() {}
```
I think it would be good to add `staged_api` as a feature gate for this so that users can't reference this lint.

It should hopefully work by just adding a line similar to https://github.com/rust-lang/rust/blob/94b4de0e0793c8921d30e0fb886be712d17db6e5/compiler/rustc_session/src/lint/builtin.rs#L535

**Review by lcnr — APPROVED** [review]

**Comment by m-ou-se** [comment]
> Does this mean that the following test compiles rn?

Yes. Every `rustc::` or `whatever::` lint is ignored when unknown.

All other internal lints also compile just fine like that:
```rust
#![allow(rustc::usage_of_qualified_ty)]
fn main() {}
```

Tool lints (with `::`) don't support `@feature_gate`. If I turn it into a regular lint, it'll show up in `-W help` even though it should be internal only, and the annotations I added for `Wake` will need to be changed to `#[cfg_attr(not(bootstrap), allow(...)]` to stop the bootstrap/beta rustc from complaining about it until the next release cycle.

**Comment by lcnr** [comment]
hmm, ok :+1: 

Then r=me with the merge commit removed. @petrochenkov do you still have some concerns or questions about this?

**Comment by petrochenkov** [comment]
Ok, the previous time I didn't look at the code and misinterpreted what this PR does.
This lint is more against *outdated* `#[unstable]` attributes, rather anything else.

We have some conventions regarding lint names (https://github.com/rust-lang/rfcs/blob/master/text/0344-conventions-galore.md#lints), and regarding diagnostic messages ("An `#[unstable]` annotation here ..." -> "an `#[unstable]` annotation here ...") which this lint doesn't match, but it doesn't matter much since all of this is internal and unstable anyway.

r? @lcnr

**Comment by m-ou-se** [comment]
> broken_heart Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1105725867)

Uh, that failure seems unrelated. `ui/autoref-autoderef/autoderef-and-borrow-method-receiver.rs` failed with no output and exit code 0xc0000005 (access violation). But that test is just some dead code and `fn main() {}`.

**Comment by lcnr** [comment]
yeah, let's hope this one was spurious

@bors retry

**Pull request merged**